src/debug_sdfs_ed64.c

```diff
@@ -0,0 +1,590 @@
+
+// Everdrive communication mode - see everdrive_sd_set_mode
+#define ED64_SD_MODE_CMD_READ           0x00
+#define ED64_SD_MODE_CMD_WRITE          0x04
+#define ED64_SD_MODE_DATA_READ          0x08
+#define ED64_SD_MODE_DATA_WRITE         0x0C
+
+static uint32_t everdrive_sd_active_mode = 0;
+
+// Set the mode to talk to the SD card. In ED64_SD_MODE_CMD_X modes, the
+// bytes are read from/written to the cmd line. WithED64_SD_MODE_DATA_X,
+// the provided bytes are written in SD wide bus format. e.g for [abcd efgh],
+// the output on the 4 data lanes will look like;
+// dat3: ae
+// dat2: bf
+// dat1: cg
+// dat0: dh
+// Effectively, the clock line will be toggled by the amount set by
+// set_everdrive_sd_bitlen for each read/write. For example on the data mode,
+// setting the bit len to 2 will output a single byte for every
+// everdrive_sd_write
+static void everdrive_sd_set_mode(uint8_t mode) {
+	if ((everdrive_sd_active_mode & ED64_SD_MODE_COMM) == mode) return;
+	everdrive_sd_active_mode &= ~ED64_SD_MODE_COMM;
+	everdrive_sd_active_mode |= mode;
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + mode, 0xffff);
+}
+
+static uint32_t __attribute__((aligned(16))) everdrive_sd_config = 0;
+// Sets how many bits are read/written at a time, per lane. This is equal to the
+// times the clock line is toggled per read/write. The actual data manipulated
+// depends on the mode (see everdrive_sd_set_mode)
+static void set_everdrive_sd_bitlen(uint8_t val) {
+	everdrive_sd_config &= ~ED64_SD_CFG_BITLEN;
+	everdrive_sd_config |= (val & ED64_SD_CFG_BITLEN);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_STATUS, everdrive_sd_config);
+}
+
+void everdrive_sd_busy() {
+	while ((io_read(ED64_BASE_ADDRESS + ED64_REGISTER_SD_STATUS)
+		& ED64_SD_STATUS_BUSY) != 0);
+}
+
+void everdrive_sd_write_command(uint8_t val) {
+	everdrive_sd_set_mode(ED64_SD_MODE_CMD_WRITE);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_CMD_WRITE, val);
+	everdrive_sd_busy();
+}
+
+uint8_t everdrive_sd_read_command() {
+	everdrive_sd_set_mode(ED64_SD_MODE_CMD_READ);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_CMD_READ, 0xffff);
+	everdrive_sd_busy();
+	return io_read(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_CMD_READ);
+}
+
+void everdrive_sd_write_data(uint8_t val) {
+	everdrive_sd_set_mode(ED64_SD_MODE_DATA_WRITE);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_DATA_WRITE,  0x00ff | (val << 8));
+}
+
+uint8_t everdrive_sd_read_data() {
+	everdrive_sd_set_mode(ED64_SD_MODE_DATA_READ);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_DATA_READ,  0xffff);
+	return io_read(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_DATA_READ);
+}
+
+// Wait for and read the first byte
+uint8_t everdrive_sd_read_first() {
+	uint8_t res;
+	uint32_t timeout = 64; // 8 bytes * 8 bits
+
+	set_everdrive_sd_bitlen(8);
+	res = everdrive_sd_read_command();
```

Comment:
Why here you're reading 8 bits and then checking also the MSB? It would seem to me that you should get into `bitlen(1)` immediately, or alternatively check all bits read here.

Reply:
By setting the bit length to 8 we shift a byte immediately and then we start checking one by one. This saves a few cycles and is probably more readable.

 e.g
```
xxxx xxxx 1234 5678
\_______/ \_______/
  our      new response from card
buffer

After first 8 bit shift;

1234 5678 90ab cdef
\ we check these bits

After a single shift;

2345 6789 0abc defg
\ we check these bits
```
So this way we just dispose our initial 8 bits which , we don't care.

Reply:
My doubt is that of the initial 8 bits, you actually do check the MSB only and ignore the other 7. When the loop start, `res` is the value read with bitlen(8).

Reply:
No, we don't ignore them. When the loop start, `res` is a whole byte read but each 1 bit read shifts only a single bit into `res` from the right thus in each iteration we sequentially check each bit of that word. Am I missing something here?

Reply:
My understanding is that bitlen(8) + read means "extract 8 bits from the shift register". If this understanding is correct, then first read is reading 8 bits from the queue. The loop is checking the MSB. Then in the loop you call bitlen(1) + read which I think it means "extract 1 bit from the shift register", so it will extract the first bit that follows the 8 you have just extracted. This means that of the first 8 bits, you only checked 1 bit.

In your ASCII example in the comment above, it seems like you must completely discard and ignore the first 8 bit. In that case, you probably want to either reset res before the loop, or change the loop into a `do/while` so that the condition is at the end of the loop so that immediately uses the 1-bit read.